### ISSUE TITLE
Robust handling of glass catalog changes

### ISSUE THREAD

**Issue by cdhf** [microsoft/OpticSim.jl#107]
One problem found in #57 is a discontinued glass that seems to be removed in newer versions of the AGF file downloads.

It makes sense to keep even discontinued glasses. Firstly, it can happen that you need to modify an old design to replace a discontinued glass after it has already gone, and then you don't have the old glass to simulate a baseline. Secondly, there are companies that keep stock of old glasses for special applications, so discontinued does not necessarily mean unavailable.

I can think of two ways to do this:

- Seek permission from the glass manufacturers to include the AGF files in this project. The glass catalog is then uniquely identified by the OpticSim version.
- Make easy to use utility functions to store the glass catalogs in the repository of the simulation using this package. The user can then add the files to their own version control system and decide, when/if to update a catalog.

**Comment by alfredclwong** [comment]
I've discussed the first option with @BrianGun and the general consensus is that this approach is a bit too tedious as we might need to get lawyers involved.

Because of this, I coded up a quick implementation of the second option, which is described [here](https://microsoft.github.io/OpticSim.jl/dev/glasscat/). The documentation isn't very clear (I'm actually working on a branch locally to improve the clarity) but it does support the ability to add catalogs in your local version of OpticSim. If you have any questions/suggestions for this part of the package, let me know!

**Comment by cdhf** [comment]
@alfredclwong I couldn't find anything in the documentation you linked to that seems to refer to this feature.

I also think that it makes sense to build the glass catalog at runtime explicitely. It could look like this:

```
using OpticSim

(run some OpticSim function, get an exception asking to install the glass catalogs)

setup_glasses(directory = <a standard directory in the current project>)
(downloads the files, builds and stores all in the directory so that the user can put it under version control)

or/and

add_glasses(agf_files...; directory = <a standard directory in the current project>)
(merges agf files to the glass catalog)
```
I am actually working in a company that makes optical materials as well, so in this case it is good if one has the ability to add self defined glasses.

**Comment by alfredclwong** [comment]
Just want to check that we're talking about the same thing. In sources.txt, a list of glass catalogs is defined, which are downloaded and added to the project automatically. This can be extended by any user to incorporate additional catalogs. Is this the kind of base functionality that you're talking about?

https://github.com/microsoft/OpticSim.jl/blob/999b8a64564b4441fe76643a0b98b9dfb4ed8a7d/deps/sources.txt#L1-L6

Regarding runtime vs compile-time, I remember now that this is a discussion that I had with @BrianGun previously. We build the catalogs at compile-time because this gives access to autocomplete for installed glass catalogs, in that you can type "SCHOTT." and see the available selection of glasses in an IDE.

In the general case, I have a natural inclination towards the kind of solution that you're suggesting, but @BrianGun has a strong preference for autocomplete, which necessitates our current solution. Is this accurate, Brian?

**Comment by friggog** [comment]
You could have a function which copies the agf file to the package dir and rebuilds the files at runtime making it easy for the user to add custom files. Then on future loads auto complete should still work with the new files as they will all get read from the cache the same as it is now

**Comment by BrianGun** [comment]
I agree with @friggog. So long as the AGFGlassCat.jl file gets rebuilt when new glasses are added it doesn't make any difference to me whether this is done only at precompile time or also at run time. 

The VSCode intellisense system seems to be smart enough to automatically update when files are changed regardless of whether this happens at precompilation time or run time. So you might still get good autocompletion even though the glass catalog is being updated dynamically.

In other words if being able to load glass files at run time, as well as at precompile time, is important then there doesn't seem to be a good reason not to do it. Go ahead.

**Comment by alfredclwong** [comment]
Sounds like the best of both worlds - I'll aim to get an implementation done asap. Should be up for review tomorrow.
